stl/inc/span

```diff
@@ -0,0 +1,723 @@
+#endif // _ITERATOR_DEBUG_LEVEL
+    }
+
+private:
+#if _ITERATOR_DEBUG_LEVEL >= 1
+    pointer _Mydata = nullptr;
+    size_t _Mysize  = 0;
+    size_t _Myoff   = 0;
+#else // ^^^ _ITERATOR_DEBUG_LEVEL >= 1 ^^^ // vvv _ITERATOR_DEBUG_LEVEL == 0 vvv
+    pointer _Myptr = nullptr;
+#endif // _ITERATOR_DEBUG_LEVEL
+};
+
+// VARIABLE TEMPLATE _Is_span_v
+template <class _Ty, size_t _Extent>
+class span;
+
+template <class>
+inline constexpr bool _Is_span_v = false;
+
+template <class _Ty, size_t _Extent>
+inline constexpr bool _Is_span_v<span<_Ty, _Extent>> = true;
+
+// VARIABLE TEMPLATE _Is_std_array_v
+template <class>
+inline constexpr bool _Is_std_array_v = false;
+
+template <class _Ty, size_t _Extent>
+inline constexpr bool _Is_std_array_v<array<_Ty, _Extent>> = true;
+
+// VARIABLE TEMPLATE _Is_span_convertible_range_v
+template <class _Cont, class _Ty>
+inline constexpr bool _Is_span_convertible_range_v =
+    is_convertible_v<remove_pointer_t<decltype(_STD data(_STD declval<_Cont&>()))> (*)[], _Ty (*)[]>;
+
+// VARIABLE TEMPLATE _Has_container_interface_v
+template <class, class = void>
+inline constexpr bool _Has_container_interface_v = false;
+
+template <class _Cont>
+inline constexpr bool _Has_container_interface_v<_Cont,
+    void_t<decltype(_STD data(_STD declval<_Cont>())), decltype(_STD size(_STD declval<_Cont>()))>> = true;
+
+// VARIABLE TEMPLATE _Is_span_compatible_container_v
+// clang-format off
+template <class _Cont, class _Ty>
+inline constexpr bool _Is_span_compatible_container_v =
+       !_Is_span_v<_Cont>
+    && !_Is_std_array_v<_Cont>
+    && !is_array_v<_Cont>
+    && _Is_span_convertible_range_v<_Cont, _Ty>
+    && _Has_container_interface_v<_Cont>;
+// clang-format on
+
+// [views.span]
+// CLASS TEMPLATE span
+template <class _Ty, size_t _Extent = dynamic_extent>
+class span {
+public:
+    using element_type           = _Ty;
+    using value_type             = remove_cv_t<_Ty>;
+    using index_type             = size_t;
+    using difference_type        = ptrdiff_t;
+    using pointer                = _Ty*;
+    using const_pointer          = const _Ty*;
+    using reference              = _Ty&;
+    using const_reference        = const _Ty&;
+    using iterator               = _Span_iterator<pointer>;
+    using const_iterator         = _Span_iterator<const_pointer>;
+    using reverse_iterator       = _STD reverse_iterator<iterator>;
+    using const_reverse_iterator = _STD reverse_iterator<const_iterator>;
+
+    static constexpr index_type extent = _Extent;
+
+    // [span.cons] Constructors, copy, and assignment
+    constexpr span() noexcept = default;
+
+    constexpr span(pointer _Ptr, index_type _Count) noexcept : _Mydata(_Count, _Ptr) {}
+
+    constexpr span(pointer _First, pointer _Last) noexcept : _Mydata(static_cast<index_type>(_Last - _First), _First) {
+#if _CONTAINER_DEBUG_LEVEL > 0
+        _STL_VERIFY(_First < _Last, "span construction from pointers only valid for first > last");

```

Comment:
There are two bugs here:

* It is okay for `first == last`. That's just an empty span.
* Your message says `>` which is incorrect.

Reply:
Thanks a lot